### PR TITLE
Javascript/javascript3/week3

### PR DESCRIPTION
**Pull request by cph-kiwi**
I have completed my javascript/javascript3/week3 homework.

### CODE REVIEW

File: javascript/javascript3/week3/circles/circles.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
+const bodyElement = document.querySelector("body");
+const buttonElement = document.querySelector("#button");
+const canvas = document.createElement("canvas");
+
+const context = canvas.getContext("2d");
+
+canvas.width = window.innerWidth - 100;
+canvas.height = window.innerHeight - 200;
+
+class Circle {
```
Comment (aleksd-dfds) [inline javascript/javascript3/week3/circles/circles.js:10]: it's a good practice to have classes in their own file. But of course the assignment says nothing about that...

File: javascript/javascript3/week3/classes-and-promises/promises.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+const user2 = new User("Beth", "Jackson", "cph-kiwi");
+
+const user3 = new User("Dlnya", "Mazhari", "DlnyaMazhari");
+
+const user4 = new User("Mohammed", "Enqira", "Enqira");
+
+const user5 = new User("Tithi", "Kundu", "TITHI1244");
+
+function fetchAllAtOnce() {
+  Promise.all([
```
Comment (aleksd-dfds) [inline javascript/javascript3/week3/classes-and-promises/promises.js:51]: good job on Promise.all(), looks right.

File: javascript/javascript3/week3/classwork/exercises.js [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+  }
+
+  addJob(job) {
+    this.jobs.push(job);
+  }
+
+  // filter returns a new array, whereas here I want to modify the original - hence I've used a loop and splice
+  removeJob(job) {
+    for (let i = 0; i < this.jobs.length; i++) {
+      if (this.jobs[i] === job) {
```
Comment (aleksd-dfds) [inline javascript/javascript3/week3/classwork/exercises.js:46]: you can also use "find" or "filter" here.

File: javascript/javascript3/week3/classwork/exercises.js [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    // this.jobs = this.jobs.filter((currentjob) => currentjob.id !== job.id);
+  }
+
+  addEducation(education) {
+    this.educations.push(education);
+  }
+
+  removeEducation(education) {
+    for (var i = 0; i < this.educations.length; i++) {
+      if (this.educations[i] === education) {
```
Comment (aleksd-dfds) [inline javascript/javascript3/week3/classwork/exercises.js:60]: same here

File: javascript/javascript3/week3/classwork/shapes.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
+class Shape {
```
Comment (aleksd-dfds) [inline javascript/javascript3/week3/classwork/shapes.js:1]: looks familiar.

File: javascript/javascript3/week3/shopping-cart/shopping-cart.js [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+      ulElement.appendChild(liElement);
+      deleteButton.addEventListener("click", () => {
+        shoppingCart.removeItem(productIndex);
+        totalPElement.textContent = `Shopping cart total: ${shoppingCart.getTotal()}`;
+        shoppingCart.renderProducts();
+      });
+    });
+  }
+
+  getUser() {
```
Comment (aleksd-dfds) [inline javascript/javascript3/week3/shopping-cart/shopping-cart.js:93]: yeah sure, this'll work, but async / await is better

File: javascript/javascript3/week3/shopping-cart/shopping-cart.js [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+class Product {
+  constructor(name, price) {
+    this.name = name;
+    this.price = price;
+  }
+
+  convertToCurrency(currency) {
+    if (this.rates) {
+      const rate = this.rates[currency];
+      return Promise.resolve(Math.round(this.price * rate));
```
Comment (aleksd-dfds) [inline javascript/javascript3/week3/shopping-cart/shopping-cart.js:27]: Hmm ... something is wrong here. 
1) Shouldn't this function be async?
2) Promise.resolve is not necessary, the conversion happens automatically if the function is async. 
3) Use await, as much as possible, this is very hard to read and has a higher probability of error

File: javascript/javascript3/week3/shopping-cart/shopping-cart.js [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    const url = `https://api.exchangeratesapi.io/latest?base=DKK`;
+    return fetch(url)
```
Comment (aleksd-dfds) [inline javascript/javascript3/week3/shopping-cart/shopping-cart.js:31]: this looks like my early attempts at grasping .then() chaining. Even if it works, it's hard to read or prove that it works. You always win by using async / await unless you absolutely have to use .then().

**Review by aleksd-dfds — APPROVED** [review]

File: javascript/javascript3/week3/shopping-cart/shopping-cart.js (hunk 6 above)
Comment (cph-kiwi) [reply javascript/javascript3/week3/shopping-cart/shopping-cart.js:93]: Ok, I'll try to work in async/await more in the future.

File: javascript/javascript3/week3/shopping-cart/shopping-cart.js (hunk 7 above)
Comment (cph-kiwi) [reply javascript/javascript3/week3/shopping-cart/shopping-cart.js:27]: Is it "wrong"? It works as intended. But I take your point about being more efficient and reducing probability of error.

File: javascript/javascript3/week3/shopping-cart/shopping-cart.js (hunk 8 above)
Comment (cph-kiwi) [reply javascript/javascript3/week3/shopping-cart/shopping-cart.js:31]: Good to know.

File: javascript/javascript3/week3/shopping-cart/shopping-cart.js (hunk 7 above)
Comment (aleksd-dfds) [reply javascript/javascript3/week3/shopping-cart/shopping-cart.js:27]: It's not "wrong" but it's convoluted.

**Pull request merged**